revocation/ocsp/ocsp.go

```diff
@@ -119,57 +123,54 @@ func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts O
 
 	// Validate OCSP response isn't expired
 	if time.Now().After(resp.NextUpdate) {
-		return toServerResult(server, OCSPCheckError{Err: errors.New("expired OCSP response")})
+		return toServerResult(server, GenericError{Err: errors.New("expired OCSP response")})
 	}
 
-	// Check for presence of pkix-ocsp-no-check and id-ce-invalidityDate
-	// extensions in response
-	foundNoCheck := false
-	foundInvalidityDate := false
-	var invalidityDate time.Time
-	for _, extension := range resp.Extensions {
-		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
-			foundNoCheck = true
-		}
-		if !foundInvalidityDate && extension.Id.String() == invalidityDateOID {
-			foundInvalidityDate = true
-			rest, err := asn1.UnmarshalWithParams(extension.Value, &invalidityDate, "generalized")
-			if len(rest) > 0 || err != nil {
-				// invalidity date is invalid, treating as not present
-				foundInvalidityDate = false
-			}
-		}
-	}
+	// Handle pkix-ocsp-no-check and id-ce-invalidityDate extensions if present
+	// in response
+	extensionMap := extensionsToMap(resp.Extensions)
 	//lint:ignore SA9003 // Empty body in an if branch (Remove after adding CRL)
-	if !foundNoCheck {
+	//lint:ignore SA4006 // Unused value (Remove after adding CRL)
+	if _, foundNoCheck := extensionMap[pkixNoCheckOID]; !foundNoCheck {
 		// This will be ignored until CRL is implemented
 		// If it isn't found, CRL should be used to verify the OCSP response
 		// TODO: add CRL support
 		// https://github.com/notaryproject/notation-core-go/issues/125
 	}
+	if invalidityDateBytes, foundInvalidityDate := extensionMap[invalidityDateOID]; foundInvalidityDate && !opts.SigningTime.IsZero() && resp.Status == ocsp.Revoked {
+		var invalidityDate time.Time
+		rest, err := asn1.UnmarshalWithParams(invalidityDateBytes, &invalidityDate, "generalized")
+		if len(rest) == 0 && err == nil && opts.SigningTime.Before(invalidityDate) {
+			return toServerResult(server, nil)
+		}
+	}
```

Comment:
Should we put this logic between line 152 and 153?

Reply:
This was discussed, but it was recommended to separate the invalidity logic date from the switch statement for increased readability and understanding.